sqlmodel_repository/base_repository.py

```diff
@@ -40,6 +57,8 @@ def _update(self, entity: GenericEntity, **kwargs) -> GenericEntity:
             This method must use the same context to fetch and update the entity. Otherwise its detached and may not be updated.
         """
         session = self.get_session()
+        self._emit_log("Updating", entities=[entity], **kwargs)
```

Comment:
I think the most relevant log is after the operation is actually finished. Otherwise you can't really tell if the operation succeeded just from the logs. You could argue that there will probably be an exception in case it didn't succeed, but depending on the issue this must not necessarily be the case.

The "safest" approach would be to just log twice - before and after the operation. But I guess that spams the logs a lot... Personally I prefer to have repository logs always after the operation, but I'm also fine with leaving it like this if you have another opinion.

Just an example why I think like this: Logs for an API call in our case would be something like this:

- REQUEST INCOMING
- [optional] service layer log
- repository log
- [optional] service layer log
- REQUEST FINISHED

In that case you'll likely know the context very well and by looking in the code you know what you expect. So having log that tell you what actually happened is more beneficial than logs that tell you what will happen in the future 😄

Reply:
I wasn't sure, so I asked a GPT4 based model and also forced it to use literature sources only with citing them at the end 😄 Here's the (pretty useful in my opinion) result:

There are different opinions on where to log in the repository pattern, and it largely depends on the context and specific requirements of your project. Here are some potential pros and cons of logging before and after code execution:

Logging before code execution:

Pros:
You can check that the inputs to your code are correct and as expected before any changes are made to the system.
You can see which functions or methods are called and in what order, which can be helpful for debugging and understanding the control flow of the code.
Cons:
If the inputs to the code are complex and there are many edge cases, logging before code execution can lead to a lot of log messages and cluttered logs.
If the code changes the state of the system, you might not see the actual effect of the changes until after they have been made.
Logging after code execution:

Pros:
You can see the actual effect of the code on the system, which can be more helpful than just logging inputs.
You can catch any errors or exceptions that occur during the code execution and log them.
Cons:
You might not see which functions or methods were called and in what order, which can make it harder to understand the control flow of the code.
If the code has many side effects or changes many parts of the system, logging after code execution can lead to a lot of log messages and cluttered logs.
As for a recommendation, I would suggest a combination of both. Logging before code execution can be useful to ensure the correct inputs are being passed to the code, and logging after code execution can be useful to see the actual effect of the code on the system. However, it's important to strike a balance and avoid cluttering the logs with too much information. I recommend following the principles of structured logging, which allow you to log messages in a consistent and structured way, making it easier to search, filter, and analyze logs.

Regarding literature, there are several sources that discuss logging in the context of the repository pattern and software engineering in general. Here are a few:

"Clean Code: A Handbook of Agile Software Craftsmanship" by Robert C. Martin. This book provides guidance on writing clean, maintainable, and testable code, including the use of logging and other debugging techniques.

"Logging and Log Management: The Authoritative Guide to Understanding the Concepts Surrounding Logging and Log Management" by Anton Chuvakin and Kevin J. Schmidt. This book covers the fundamentals of logging, log management, and log analysis, including best practices for logging in software systems.

"The Twelve-Factor App" by Adam Wiggins. This methodology for building software-as-a-service applications includes guidance on logging, including the use of structured logging to improve the manageability and observability of applications.

Overall, the recommendation to use a combination of logging before and after code execution is based on common best practices in software engineering and the principles of structured logging.

Reply:
@JonasScholl I can largely agree on your opinion and also this. Lets take both?

Reply:
How would such a optimal log look like for your?

1. Invoke
2. Log constraints 
3. Process
4. ?

Reply:
ah sure asking a GPT4 model is usegul 😂 yes then let's just take both

Reply:
> How would such a optimal log look like for your?
> 
> 
> 
> 1. Invoke
> 
> 2. Log constraints 
> 
> 3. Process
> 
> 4. ?

not sure what you mean with this, the data that is logged with one entry or the log entries for a repository method call?

Reply:
I mean at the beginning we log the constraints of the request (kwargs). What would we log after the operation completed successfully?

Reply:
ah okay, I guess logging the success of the operation with the entity ID(s) would suffice - at least if we're providing the other information before